ocaml/gencert/validation.ml

```diff
@@ -0,0 +1,247 @@
+(*
+ * Copyright (C) Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+module Net = Network_client.Client
+
+module D = Debug.Make (struct let name = "gencert_validation" end)
+
+exception Unexpected_address_type of string
+
+open Api_errors
+open Rresult
+
+let supported_cert_hashes = [`SHA256]
+
+let validate_private_key pkcs8_private_key =
+  let ensure_key_length = function
+    | `RSA priv ->
+        let length = Mirage_crypto_pk.Rsa.priv_bits priv in
+        if length < 2048 || length > 4096 then
+          Error
+            (`Msg
+              ( server_certificate_key_rsa_length_not_supported
+              , [Int.to_string length] ))
+        else
+          Ok (`RSA priv)
+  in
+  let raw_pem = Cstruct.of_string pkcs8_private_key in
+  X509.Private_key.decode_pem raw_pem
+  |> R.reword_error (fun (`Msg err_msg) ->
+         let unknown_algorithm = "Unknown algorithm " in
+         if Astring.String.is_prefix ~affix:"multi-prime RSA" err_msg then
+           `Msg (server_certificate_key_rsa_multi_not_supported, [])
+         else if Astring.String.is_prefix ~affix:unknown_algorithm err_msg then
+           `Msg
+             ( server_certificate_key_algorithm_not_supported
+             , [
+                 Astring.String.with_range
+                   ~first:(String.length unknown_algorithm)
+                   err_msg
+               ] )
+         else (
+           D.info {|Failed to validate private key because "%s"|} err_msg ;
+           `Msg (server_certificate_key_invalid, [])
+         ))
+  >>= ensure_key_length
+
+let has_valid_time time certificate =
+  let not_before, not_after = X509.Certificate.validity certificate in
+  Ptime.(is_later ~than:not_before time && is_earlier ~than:not_after time)
+
+let load_server_certificate pem =
```

Comment:
This is not a good name as it implies loading from the file system. This is just decoding a PEM.

Reply:
I've changed the function that simply decode pem formats and make some sanity checks to `decode_...` to avoid confusion.